tls/s2n_x509_validator.c

```diff
@@ -789,12 +795,16 @@ S2N_RESULT s2n_x509_validator_validate_cert_chain(struct s2n_x509_validator *val
         RESULT_GUARD_POSIX(s2n_extension_list_process(S2N_EXTENSION_LIST_CERTIFICATE, conn, &first_certificate_extensions));
     }
 
-    if (conn->config->cert_validation_cb) {
-        struct s2n_cert_validation_info info = { 0 };
-        RESULT_ENSURE(conn->config->cert_validation_cb(conn, &info, conn->config->cert_validation_ctx) >= S2N_SUCCESS,
+    /* skip this step on the re-entry case */
+    if (validator->state != AWAITING_VALIDATE_CALLBACK && conn->config->cert_validation_cb) {
+        RESULT_ENSURE(conn->config->cert_validation_cb(conn, &(validator->info), conn->config->cert_validation_ctx) >= S2N_SUCCESS,
                 S2N_ERR_CANCELLED);
-        RESULT_ENSURE(info.finished, S2N_ERR_INVALID_STATE);
-        RESULT_ENSURE(info.accepted, S2N_ERR_CERT_REJECTED);
+        RESULT_GUARD(s2n_handle_cert_validation_callback_result(validator));
+    }
+
+    /* update state after completing the async validation */
+    if (validator->state == AWAITING_VALIDATE_CALLBACK) {
+        validator->state = VALIDATED;
     }
```

Comment:
Currently it's possible for the validator state to remain set to `INIT` if cert validation is disabled: 
https://github.com/aws/s2n-tls/blob/e4a5a74b0492767ac996e93a7257d533087015dd/tls/s2n_x509_validator.c#L550-L557

To avoid affecting the assumptions of other parts of code relying on `validator->state`, we may want to ensure the `INIT` state continues to represent a handshake with no x509 validation performed by s2n-tls.